### PR TITLE
Feature/integrate sora crypto

### PR DESCRIPTION
**Pull request by Warchant**
In this PR I integrated soramitsu/sora-crypto-java into DID-Resolver

**Note**: Initially, I thought I need to use `.sign(Object object, ...)` interface, where Object can be POJO, JsonNode, Map... But after integration (and struggling with ObjectMapper) I think it is better to accept serialized JSON (String) OR JsonNode as input argument, then sign it and return signed JsonNode. What do you think?

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -57,7 +57,7 @@ ext {
     jUnitVersion = '5.2.0'
     jacksonVersion = '2.9.6'
     jUnitPlatformVersion = '1.2.0'
-    ed25519Version = '1.3.3'
+    soraCryptoVersion = '8e4ace36a1'
```
Comment (rogachevsn55) [inline build.gradle]: nice version of library) what will be next?)

File: src/main/java/jp/co/soramitsu/sora/didresolver/dto/Authentication.java [hunk 2]
```diff
@@ -1,8 +1,8 @@
 package jp.co.soramitsu.sora.didresolver.dto;
 
-import java.net.URI;
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
 import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import java.net.URI;
```
Comment (rogachevsn55) [inline src/main/java/jp/co/soramitsu/sora/didresolver/dto/Authentication.java]: Why did you make this change?

File: src/main/java/jp/co/soramitsu/sora/didresolver/dto/PublicKey.java [hunk 3]
```diff
@@ -1,8 +1,8 @@
 package jp.co.soramitsu.sora.didresolver.dto;
 
-import java.net.URI;
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
 import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import java.net.URI;
```
Comment (rogachevsn55) [inline src/main/java/jp/co/soramitsu/sora/didresolver/dto/PublicKey.java]: Why did you make this change?

File: src/main/java/jp/co/soramitsu/sora/didresolver/services/impl/VerifyServiceImpl.java [hunk 4]
```diff
@@ -21,27 +27,44 @@
 @Slf4j
 public class VerifyServiceImpl implements VerifyService {
 
+  private static final ObjectMapper mapper = new ObjectMapper()
```
Comment (rogachevsn55) [inline src/main/java/jp/co/soramitsu/sora/didresolver/services/impl/VerifyServiceImpl.java]: use @FieldDefaults

File: src/main/java/jp/co/soramitsu/sora/didresolver/services/impl/VerifyServiceImpl.java [hunk 5]
```diff
@@ -21,27 +27,44 @@
...
+      .registerModule(new JavaTimeModule())
+      .enable(SerializationFeature.INDENT_OUTPUT)
```
Comment (rogachevsn55) [inline src/main/java/jp/co/soramitsu/sora/didresolver/services/impl/VerifyServiceImpl.java]: use static import at this line and line below

File: src/test/java/jp/co/soramitsu/sora/util/DataProvider.java [hunk 6]
```diff
@@ -62,12 +61,18 @@ public Proof getProofForTest() {
...
   public DDO getDDOFromJson(String fileName) throws IOException {
-    ObjectMapper objectMapper = new ObjectMapper();
-    objectMapper.registerModule(new JavaTimeModule());
-    Reader jsonReader = new BufferedReader(
-        new InputStreamReader(
-            DataProvider.class.getClassLoader().getResourceAsStream(fileName)));
-    return objectMapper.readValue(jsonReader, DDO.class);
+    ObjectMapper objectMapper = new ObjectMapper()
+        .registerModule(new JavaTimeModule())
+        .enable(SerializationFeature.INDENT_OUTPUT)
```
Comment (rogachevsn55) [inline src/test/java/jp/co/soramitsu/sora/util/DataProvider.java]: use static import at this line and line below

File: src/test/java/jp/co/soramitsu/sora/util/SignatureGenerator.java [hunk 7]
```diff
@@ -2,50 +2,76 @@
...
-    Reader jsonReader = new BufferedReader(
-        new InputStreamReader(
-            SignatureGenerator.class.getClassLoader().getResourceAsStream("ddo.json")));
-    DDO ddo = objectMapper.readValue(jsonReader, DDO.class);
-    KeyPair keyPair = new Ed25519Sha3().generateKeypair();
+      throws IOException, SignatureException, NoSuchAlgorithmException {
+
+    ObjectMapper objectMapper = new ObjectMapper()
+        .registerModule(new JavaTimeModule())
+        .enable(SerializationFeature.INDENT_OUTPUT)
```
Comment (rogachevsn55) [inline src/test/java/jp/co/soramitsu/sora/util/SignatureGenerator.java]: use static import at this line and line below

File: src/test/java/jp/co/soramitsu/sora/util/SignatureGenerator.java [hunk 8]
```diff
@@ -2,50 +2,76 @@
...
+        .setSerializationInclusion(Include.NON_NULL);
+
+    JSONEd25519Sha3SignatureSuite documentSignatureService = new JSONEd25519Sha3SignatureSuite(
+        new SecurityProvider(), new JSONCanonizerWithOneCoding(), objectMapper);
+
+    InputStream in = SignatureGenerator.class.getClassLoader().getResourceAsStream("ddo.json");
+    DDO ddo = objectMapper.readValue(in, DDO.class);
+
+    KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance(EdDSAKey.KEY_ALGORITHM);
```
Comment (rogachevsn55) [inline src/test/java/jp/co/soramitsu/sora/util/SignatureGenerator.java]: use static import

File: build.gradle (hunk 1 above)
Comment (Warchant) [reply build.gradle]: It is simply commit version. I can create a release and put tag here, actually.

File: src/main/java/jp/co/soramitsu/sora/didresolver/dto/Authentication.java (hunk 2 above)
Comment (Warchant) [reply src/main/java/jp/co/soramitsu/sora/didresolver/dto/Authentication.java]: IDE did this

File: src/main/java/jp/co/soramitsu/sora/didresolver/dto/PublicKey.java (hunk 3 above)
Comment (Warchant) [reply src/main/java/jp/co/soramitsu/sora/didresolver/dto/PublicKey.java]: IDE did this

File: src/main/java/jp/co/soramitsu/sora/didresolver/services/impl/VerifyServiceImpl.java (hunk 4 above)
Comment (Warchant) [reply src/main/java/jp/co/soramitsu/sora/didresolver/services/impl/VerifyServiceImpl.java]: field is static

File: src/main/java/jp/co/soramitsu/sora/didresolver/services/impl/VerifyServiceImpl.java (hunk 5 above)
Comment (Warchant) [reply src/main/java/jp/co/soramitsu/sora/didresolver/services/impl/VerifyServiceImpl.java]: Sorry, how? 
Do you mean that I need to define this mapper somewhere and then import it?

File: src/test/java/jp/co/soramitsu/sora/util/SignatureGenerator.java (hunk 8 above)
Comment (Warchant) [reply src/test/java/jp/co/soramitsu/sora/util/SignatureGenerator.java]: Sorry, I did not get this

**Comment by tdorzhi** [comment]
>I think it is better to accept serialized JSON (String) OR JsonNode as input argument, then sign it and return signed JsonNode. What do you think?

I think it'd be better that sign accepts either HEX64 string (that will be transformed into byte array), or byte array itself, resolver should manage how to get the blob from input data

File: src/test/java/jp/co/soramitsu/sora/util/SignatureGenerator.java [hunk 9]
```diff
@@ -1,51 +1,78 @@
...
+        proof.getCreated().toString(),
+        proof.getCreator().toString(),
+        proof.getNonce(),
+        proof.getPurpose()
+    );
+
+    val signed = documentSignatureService
+        .sign(ddo, (EdDSAPrivateKey) keyPair.getPrivate(), options);
+
+    assert documentSignatureService.verify(signed, (EdDSAPublicKey) keyPair.getPublic());
```
Comment (tdorzhi) [inline src/test/java/jp/co/soramitsu/sora/util/SignatureGenerator.java]: Also, don't rely on this, assertions IIRC can be simply switched off at JVM startup

Comment (Warchant) [reply src/test/java/jp/co/soramitsu/sora/util/SignatureGenerator.java]: I know. This assert is just self-check.

**Pull request merged**